src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -33,16 +33,28 @@ object JoinIndexRanker {
    * number of buckets, better the parallelism achieved during join in general, assuming there is
    * no resource constraint.
    *
+   * If hybridScanEnabled is true, pick the index pair which covers the most source data
+   * so that we could minimize the amount of data for on-the-fly shuffle or merge.
+   *
    * @param indexPairs index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
+   * @param hybridScanEnabled HybridScan config.
    * @return rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
-      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
+      indexPairs: Seq[(IndexLogEntry, IndexLogEntry)],
+      hybridScanEnabled: Boolean = false): Seq[(IndexLogEntry, IndexLogEntry)] = {
 
     indexPairs.sortWith {
       case ((left1, left2), (right1, right2)) =>
-        if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
+        if (hybridScanEnabled && ((left1.allSourceFileInfos.size + left2.allSourceFileInfos.size) >
+              (right1.allSourceFileInfos.size + right2.allSourceFileInfos.size))) {
+          true
+        } else if (hybridScanEnabled &&
```

Comment:
Add tests for these changes? Frankly, this should have been done as a separate PR so that we can iterate quicker: https://github.com/google/eng-practices/blob/master/review/developer/small-cls.md#why-write-small-cls-why